File: app/src/main/java/com/leonov_dev/currencyconverter/MainActivity.java
package com.leonov_dev.currencyconverter;

import androidx.lifecycle.ViewModelProviders;
import android.content.Intent;
import androidx.annotation.NonNull;
import com.google.android.material.navigation.NavigationView;
import androidx.fragment.app.FragmentActivity;
import androidx.fragment.app.FragmentManager;
import androidx.core.view.GravityCompat;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentTransaction;
import androidx.drawerlayout.widget.DrawerLayout;
import androidx.appcompat.app.ActionBarDrawerToggle;
import androidx.appcompat.app.AppCompatActivity;
import android.os.Bundle;
import androidx.appcompat.widget.Toolbar;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;

import com.leonov_dev.currencyconverter.preferences.CurrencySettingsActivity;


public class MainActivity extends AppCompatActivity implements NavigationView.OnNavigationItemSelectedListener {

    //TODO get rid of this cludge
    public static boolean isFirstRun = true;

    private final String LOG_TAG = this.getClass().getSimpleName();

    private FragmentTransaction ft;

    private StockConverterViewModel viewModel;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main_wrapper);

        Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);

        DrawerLayout drawerLayout = (DrawerLayout) findViewById(R.id.drawer_layout);
        ActionBarDrawerToggle toggle = new ActionBarDrawerToggle(this, drawerLayout,
                toolbar, R.string.navigation_drawer_open, R.string.navigation_drawer_close);
        drawerLayout.addDrawerListener(toggle);
        toggle.syncState();

        NavigationView navigationView = (NavigationView) findViewById(R.id.nav_view);
        navigationView.setNavigationItemSelectedListener(this);

        Fragment stockAndConverterFragment = new StockAndConverterFragment();

        ft = getSupportFragmentManager().beginTransaction();
        ft.replace(R.id.container, stockAndConverterFragment);
        ft.addToBackStack(null);
        ft.commit();

        viewModel = obtainViewModel(this);

    }

    public static StockConverterViewModel obtainViewModel(FragmentActivity activity) {
        // Use a Factory to inject dependencies into the ViewModel
        ViewModelFactory factory = ViewModelFactory.getInstance(activity.getApplication());
        StockConverterViewModel viewModel =
                ViewModelProviders.of(activity, factory).get(StockConverterViewModel.class);

        return viewModel;
    }

    @Override
    public void onBackPressed() {
        DrawerLayout drawer = (DrawerLayout) findViewById(R.id.drawer_layout);
        if (drawer.isDrawerOpen(GravityCompat.START)) {
            drawer.closeDrawer(GravityCompat.START);
        } else {
            super.onBackPressed();
        }

        FragmentManager fragmentManager = getSupportFragmentManager();
        if (fragmentManager.getBackStackEntryCount() == 1){

        }
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_top, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int elementId = item.getItemId();
        if (elementId == R.id.action_settings){
            Intent intent = new Intent(this, CurrencySettingsActivity.class);
            startActivity(intent);
            return true;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public boolean onNavigationItemSelected(@NonNull MenuItem item) {

        Class fragmentClass = null;
        Fragment fragment = null;

        int id = item.getItemId();
        if (id == R.id.home_page_item){
            fragmentClass = StockAndConverterFragment.class;
        } else if (id == R.id.about_item) {
            fragmentClass = AboutFragment.class;
        }

        try {
            fragment = (Fragment) fragmentClass.newInstance();
        } catch (Exception e) {
            Log.e(LOG_TAG, "Error Creating fragment " + e);
        }

        FragmentTransaction transaction = getSupportFragmentManager().beginTransaction();
        try {
            transaction.replace(R.id.container, fragment);
            transaction.addToBackStack(null);
            transaction.commit();
        }catch (Exception e){
            Log.e(LOG_TAG, "Transaction error" + e);
        }

        setTitle(item.getTitle());
        DrawerLayout drawer = (DrawerLayout) findViewById(R.id.drawer_layout);
        drawer.closeDrawer(GravityCompat.START);
        return true;

    }
}

File: app/src/main/java/com/leonov_dev/currencyconverter/ui/home/MainActivity.kt
package com.leonov_dev.currencyconverter.ui.home

import android.graphics.Color
import androidx.lifecycle.ViewModelProviders
import androidx.fragment.app.FragmentActivity
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.view.MenuItem
import com.google.android.material.bottomnavigation.BottomNavigationView

import com.leonov_dev.currencyconverter.R
import com.leonov_dev.currencyconverter.ViewModelFactory
import com.leonov_dev.currencyconverter.ui.about.AboutFragment
import com.leonov_dev.currencyconverter.ui.home.converter.ConverterFragment
import com.leonov_dev.currencyconverter.ui.home.stock.StockFragment
import kotlinx.android.synthetic.main.activity_main_wrapper.*


class MainActivity : AppCompatActivity(), BottomNavigationView.OnNavigationItemSelectedListener {

    companion object {

        private val TAG_TO_FRAGMENT_FACTORY_FUNCTION = mapOf(
                ConverterFragment::class.java.simpleName to { ConverterFragment() },
                StockFragment::class.java.simpleName to { StockFragment() },
                AboutFragment::class.java.simpleName to { AboutFragment() }
        )

        fun obtainViewModel(activity: FragmentActivity): StockConverterViewModel {
            // Use a Factory to inject dependencies into the ViewModel
            val factory = ViewModelFactory.getInstance(activity.application)
            return ViewModelProviders.of(activity, factory).get(StockConverterViewModel::class.java)
        }
    }

    private var viewModel: StockConverterViewModel? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main_wrapper)

        bottomNavigationView.setOnNavigationItemSelectedListener(this::onNavigationItemSelected)
        viewModel = obtainViewModel(this)
    }

    // Handles bottom navigation bar for android
    override fun onNavigationItemSelected(menuItem: MenuItem): Boolean {
        when (menuItem.itemId) {
            R.id.navigation_rates -> presentFragment(StockFragment::class.java.simpleName)
            R.id.navigation_converter -> presentFragment(ConverterFragment::class.java.simpleName)
            R.id.navigation_settings -> presentFragment(AboutFragment::class.java.simpleName)
        }
        return true
    }

    // Handles whether to add a new fragment to the fragment container or to show an existing one
    private fun presentFragment(tag: String) {
        require(TAG_TO_FRAGMENT_FACTORY_FUNCTION.containsKey(tag)) { "Unexpected fragment tag" }

        val transaction = supportFragmentManager.beginTransaction()

        // Hide fragments that are not the one being shown.
        TAG_TO_FRAGMENT_FACTORY_FUNCTION.keys
                .filter { it != tag }
                .forEach { it ->
                    supportFragmentManager.findFragmentByTag(it)?.let { transaction.hide(it) }
                }

        // Show or create the desired fragment.
        supportFragmentManager.findFragmentByTag(tag).let {
            if (it == null) {
                val fragment = TAG_TO_FRAGMENT_FACTORY_FUNCTION[tag]!!()
                transaction.add(R.id.fragmentContainer, fragment, tag)
            } else {
                transaction.show(it)
            }
        }
        transaction.commit()
    }
}
